samples/triangle/main.c

```diff
@@ -0,0 +1,281 @@
+    out[3][2] = z_min;
+}
+
+/* Load the shader we will render with */
+static void init_shader(void)
+{
+    uint32_t *p;
+    int       i;
+
+    /* Setup vertex shader */
+    uint32_t vs_program[] = {
+        #include "vs.inl"
+    };
+
+    p = pb_begin();
+
+    /* Set run address of shader */
+    pb_push(p++, NV097_SET_TRANSFORM_PROGRAM_START, 1); *(p++)=0;
+
+    /* Set execution mode */
+    pb_push(p++, NV097_SET_TRANSFORM_EXECUTION_MODE, 2);
+    *(p++) = SHADER_TYPE_EXTERNAL;
+    *(p++) = SHADER_SUBTYPE_REGULAR;
+
+    /* Set cursor and begin copying program */
+    pb_push(p++, NV097_SET_TRANSFORM_PROGRAM_LOAD, 1); *(p++)=0;
+    for (i=0; i<sizeof(vs_program)/8; i++) {
+        pb_push(p++,NV097_SET_TRANSFORM_PROGRAM,4);
+        memcpy(p, &vs_program[i*4], 4*4);
+        p+=4;
+    }
+    pb_end(p);
+
+    /* Setup fragment shader */
+    p = pb_begin();
+    #include "ps.inl"
+    pb_end(p);
+}
+
+/* Set an attribute pointer */
+static void set_attrib_pointer(unsigned int index, unsigned int format, unsigned int size, unsigned int stride, const void* data)
+{
+    uint32_t *p = pb_begin();
+    pb_push1(p, NV097_SET_VERTEX_DATA_ARRAY_FORMAT + index*4,
+        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE, format) | \
+        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_SIZE, size) | \
+        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_STRIDE, stride));
+    p += 2;
+    pb_push1(p, NV097_SET_VERTEX_DATA_ARRAY_OFFSET + index*4, (uint32_t)data & 0x03ffffff);
+    p += 2;
+    pb_end(p);
+}
+
+/* Send draw commands for the triangles */
+static void draw_arrays(unsigned int mode, int start, int count)
+{
+    static int overflow_estimation=0;
+    #define MAX_VERTICES_PER_BATCH 65535
+    int num_batches=(count+MAX_VERTICES_PER_BATCH-1)/MAX_VERTICES_PER_BATCH;
+    uint32_t offset=0;
+    while(num_batches) {
+        int n;
+        if (num_batches > 1) {
+            n = MAX_VERTICES_PER_BATCH;
+        } else {
+            n = count % MAX_VERTICES_PER_BATCH;
+        }
+
+        uint32_t *p = pb_begin();
+        pb_push1(p, NV097_SET_BEGIN_END, mode); p += 2;
+
+        uint32_t size=(n+251)/252;
+        pb_push(p++,0x40000000|NV097_DRAW_ARRAYS,size); //bit 30 means all params go to same register 0x1810
+        
+        while(size) {
+            uint32_t num_vertices_this_batch;
+            if (size>1) 
+                num_vertices_this_batch=252; 
+            else 
+                num_vertices_this_batch=n%252;
+            
+            *(p++)=((num_vertices_this_batch-1)<<24)|offset;
+            
+            offset+=252;
+            size--;
+        }
+
+        pb_push1(p,NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END); p += 2;
+        pb_end(p);
+
+        overflow_estimation+=n;
+        if (overflow_estimation>252*8*16000) //depends on push buffer size
+        {
+            overflow_estimation=0;
+            pb_reset(); //jump back to push buffer head to avoid buffer overflow
+        }
```

Comment:
This construct looks pretty awkward. Can't we just reset more often or after a couple of iterations?
If not: Shouldn't pbkit handle this for us somehow?